lib/date.js

```diff
@@ -119,35 +119,53 @@ function govukTime (string) {
 }
 
 /**
- * Convert decorated `govukDateInput` values to an ISO 8601 date.
- *
- * The `decorate()` method applied to a `govukDateInput` creates an object with
- * `day`, `month` and `year` values. We can convert this into ISO 8601 format.
- *
- * @example <caption>Full date</caption>
- * isoDateFromDateInput({ day: '17', month: '08', year: '2021' }) // 2021-08-17
- *
- * @example <caption>Month and year only</caption>
- * isoDateFromDateInput({ month: '08', year: '2021' }) // 2021-08
- *
- * @param {object} object - Date object
- * @returns {string} `object` converted to a ISO 8601 date string
+ * Convert `govukDateInput` values to an ISO 8601 date.
+ *
+ * The `govukDateInput` creates separate values for its component for `day`,
+ * `month` and `year` values, optionally prefixed with a `namePrefix`.
+ *
+ * `namePrefix` is optional, and intended for the simplistic use case where
+ * date values are saved at the top level of prototype session data.
+ *
+ * If no `namePrefix` is provided, assumes author is setting custom names for
+ * each component value and storing session data in a nested object.
+ *
+ * @example <caption>With namePrefix</caption>
+ * data = {
+ *   'dob-day': '01',
+ *   'dob-month': '02',
+ *   'dob-year: '2012'
+ * }
+ * isoDateFromDateInput(data, 'dob') // 2012-02-01
+ *
+ * @example <caption>Without namePrefix, month and year only</caption>
+ * data = {
+ *   issued: {
+ *     month: '02',
+ *     year: '2012'
+ *   }
+ * }
+ * isoDateFromDateInput(data.issued) // 2012-02
+ *
+ * @param {object} object - Object containing date values
+ * @param {string} [namePrefix] - `namePrefix` used for date values
+ * @returns {string} ISO 8601 date string
  */
-function isoDateFromDateInput (object) {
-  object = normalize(object, '')
-
-  if (typeof object !== 'object') {
-    return object
+function isoDateFromDateInput (object, namePrefix) {
+  let day = Number(object?.day)
+  let month = Number(object?.month)
+  let year = Number(object?.year) || new Date().getFullYear()
+
+  if (namePrefix) {
+    day = Number(object[`${namePrefix}-day`])
+    month = Number(object[`${namePrefix}-month`])
+    year = Number(object[`${namePrefix}-year`])
   }
```

Comment:
Would this be clearer as an if/else statement, rather than redefining the variables if the namePrefix is set?

```suggestion
  let day, month, year
  
  if (namePrefix) {
    day = Number(object[`${namePrefix}-day`])
    month = Number(object[`${namePrefix}-month`])
    year = Number(object[`${namePrefix}-year`])
  } else {
    day = Number(object?.day)
    month = Number(object?.month)
    year = Number(object?.year) || new Date().getFullYear()  
  }
```

Also not sure why the year defaults to the current year?